pyvista/core/utilities/transformations.py

```diff
@@ -392,3 +394,180 @@ def rotation(
+
+    - When ``allow_negative_scale`` is ``True``, reflections are included with the
+      scaling ``S``. The first scaling factor is negative if there is a reflection
+      in the decomposition, and positive if there is no reflection. The y and z
+      scaling factors are always positive.
+
+    By default, compact representations of the transformations are returned (e.g.
+    as vectors or 3x3 matrix). Optionally, 4x4 matrices may be returned instead.
+
+    .. note::
+
+        The transformation is not unique.
+
+    Parameters
+    ----------
+    transformation : TransformLike
+        Array or transform to decompose.
+
+    as_matrix : bool, default: False
+        If ``True``, return translation, rotation, scaling, and shear components as
+        4x4 matrices.
+
+    allow_negative_scale : bool, default: False
+        If ``True``, the first scaling term may be negative. By default, the
+        scaling factors are always positive.
+
+    Returns
+    -------
+    numpy.ndarray
+        Length-3 translation vector (or 4x4 translation matrix if ``as_matrix`` is ``True``).
+
+    numpy.ndarray
+        3x3 rotation matrix (or 4x4 rotation matrix if ``as_matrix`` is ``True``).
+
+    numpy.ndarray
+        Length-3 scaling vector (or 4x4 scaling matrix if ``as_matrix`` is ``True``).
+
+    numpy.ndarray
+        Length-3 shear vector (or 4x4 shear matrix if ``as_matrix`` is ``True``).
+        If a vector, the values are the ``xy``, ``xz``, and ``yz`` shears that
+        fill the upper triangle above the diagonal of the shear matrix.
+
+    Examples
+    --------
+    Decompose a transformation matrix.
+
+    >>> import pyvista as pv
+    >>> matrix = [
+    ...     [0.0, -2.0, 0.0, 4.0],
+    ...     [1.0, 0.25, 0.0, 5.0],
+    ...     [0.0, 0.0, 3.0, 6.0],
+    ...     [0.0, 0.0, 0.0, 1.0],
+    ... ]
+
+    >>> T, R, S, K = pv.transformations.decompose(matrix)
+    >>> K  # shear
+    array([0.25, 0.  , 0.  ])
+
+    >>> S  # scale
+    array([1., 2., 3.])
+
+    >>> R  # rotation
+    array([[ 0., -1.,  0.],
+           [ 1.,  0.,  0.],
+           [ 0.,  0.,  1.]])
+
+    >>> T  # translation
+    array([4., 5., 6.])
+
+    """
+    matrix = _validation.validate_transform4x4(transformation)
+
+    T, R, S, K = _decompose(matrix, allow_negative_scale=allow_negative_scale)
+    if as_matrix:
+        T4 = np.eye(4, dtype=matrix.dtype)
+        T4[:3, 3] = T
+
+        R4 = np.eye(4, dtype=matrix.dtype)
+        R4[:3, :3] = R
+
+        S4 = np.diag((*S, 1)).astype(matrix.dtype)
+
+        K4 = np.eye(4, dtype=matrix.dtype)
+        K4[0, 1] = K[0]
+        K4[0, 2] = K[1]
+        K4[1, 2] = K[2]
+        return T4, R4, S4, K4
+
+    return T, R, S, K
+
+
+def _decompose(matrix, allow_negative_scale=False):
+    """Decompose a matrix into its parts.
+
+    Copied from:
+    https://github.com/matthew-brett/transforms3d/blob/6a43a98e3659d198ff6ce2c90d52ddef50fcf770/transforms3d/affines.py#L156
```

Comment:
Not sure if this is allowed or if it poses a copyright/licensing issue. Repo uses BSD-2 https://github.com/matthew-brett/transforms3d?tab=License-1-ov-file

Probably best not to add `transforms3d` as another dependency either...

Reply:
Could you explain why it is better not to add `transforms3d` as a new dependency?

Reply:
I _could_ try to use polar decomposition again from https://github.com/pyvista/pyvista/pull/6628/commits/1b66c0ae0ca2ba42ef6e24fddfc33443ca5c5412 with this code:
``` python

def _polar_decomposition(a):
      # Decompose `a=up` where u is orthonormal and p is positive semi-definite
      # See scipy.linalg.polar for details
      w, s, vh = np.linalg.svd(a, full_matrices=False)
      u = w.dot(vh)
      p = (vh.T.conj() * s).dot(vh)
      return u, p
```
Where `u` is the rotation `R`, and `p` is scaling and shear `SK`. I had problems separating `SK` earlier but they may be separable. The problem is that this is also copying code, but from `scipy` library instead of `transforms3d`. And, polar decomp uses svd, which I think makes it more computationally expensive than the other method (though I haven't tested this).

Reply:
> Could you explain why it is better not to add `transforms3d` as a new dependency?

One reason is just based on avoiding "yet another dependency". We are only using a single function that only depends on numpy, so why add the entire package?

Another reason may be testing/maintenance, their [home page](https://github.com/matthew-brett/transforms3d) states: "We use Github actions to test the code automatically under Pythons 3.7 through 3.10." So it's not officially tested for later python. But, the maintainer has pushed recent commits, and we have code coverage for the key function with our testing, so maybe it's OK.

Reply:
But, if we do keep it as a dependency, the package does have other transformation methods that could be useful to wrap in a PyVista friendly way. E.g. some of the shear methods: http://matthew-brett.github.io/transforms3d/reference/transforms3d.shears.html#transforms3d.shears.sadn2aff

E.g. similar to `Transform.rotate_vector`, we could add `Transform.shear_plane` method or something like that. I don't really use shear so not familiar with its use cases but maybe that is good to have.

Reply:
I see. How about installing it optionally, for example `pip install pyvista[transford]`. That way, there is no default Python version support or impact on load times.

Reply:
> I see. How about installing it optionally, for example `pip install pyvista[transform]`?

This is a good suggestion, however this PR is supporting #6616. And so the decomposition is needed to transform `ImageData` here:
https://github.com/pyvista/pyvista/blob/f921caf110f9d633e8784723f0ebfa90bf78e131/pyvista/core/grid.py#L921-L933

As noted in https://github.com/pyvista/pyvista/pull/6616#discussion_r1755860304, however, we don't _need_ to do a formal decomposition. We could simply estimate scale and rotation by normalizing the 3x3 part of the transformation.

It's probably best to only support one or the other, i.e. either always do formal decomposition (with `transforms3d`) as a dependency), or always do informal decomp and simply estimate it. Otherwise, different transformations would be applied based on wether the optional package is installed or not.

Reply:
It is better to always use the same calculation with and without options. It would be better to output an error message if this calculation is called, and transforms3d is not installed. Please see options such as jupyter.

Reply:
After further testing I found that this third party decomposition with `transfoms3d` package did not correctly recover some values, e.g. using a matrix with scaling (1,2,3) returned (2,1,3) in cases with a rotation (i.e. the decomposed scale was dependent on the rotation).

I have reverted to using polar decomposition, initially used earlier in https://github.com/pyvista/pyvista/pull/6628/commits/1b66c0ae0ca2ba42ef6e24fddfc33443ca5c5412.

As such, additional dependencies are no longer required.